File: ceylon/src/test/java/com/laamella/code_state_machine/PriorityTests.java
package com.laamella.code_state_machine;

import static com.laamella.code_state_machine.util.SimpleState.A;
import static com.laamella.code_state_machine.util.SimpleState.B;
import static org.junit.Assert.assertEquals;

import org.junit.Before;
import org.junit.Test;

import com.laamella.code_state_machine.builder.DslStateMachineBuilder;
import com.laamella.code_state_machine.priority.Priority;
import com.laamella.code_state_machine.util.SimpleState;

public class PriorityTests {
	private final StringBuffer trace = new StringBuffer();

	private StateMachine<SimpleState, Object, Priority> machine;

	@Before
	public void before() {
		machine = new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			@Override
			protected void executeBuildInstructions() {
				state(A).isAStartState();
				state(B).isAnEndState();
			}
		}.build();
	}

	private TraceAction trace(final String signature) {
		return new TraceAction(trace, signature);
	}

	@Test
	public void highPrioIsTheOnlyOneFiring() {
		new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			@Override
			protected void executeBuildInstructions() {
				state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
				state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));
				state(A).when(always()).transition(B, always(), Priority.LOWEST, trace("L"));
			}
		}.build(machine);

		machine.poll();
		assertEquals("H", trace.toString());
	}

	@Test
	public void normalPriosAreTheOnlyOnesFiringBecauseOtherPrioDoesntMeetCondition() {
		new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			@Override
			protected void executeBuildInstructions() {
				state(A).when(always()).transition(B, never(), Priority.HIGH, trace("H"));
				state(A).when(always()).transition(B, never(), Priority.NORMAL, trace("N"));
				state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));
				state(A).when(always()).transition(B, always(), Priority.NORMAL, trace("N"));
			}
		}.build(machine);

		machine.poll();
		assertEquals("NN", trace.toString());
	}

	@Test
	public void equalPriosFireTogether() {
		new DslStateMachineBuilder<SimpleState, Object, Priority>(Priority.NORMAL) {
			@Override
			protected void executeBuildInstructions() {
				state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
				state(A).when(always()).transition(B, always(), Priority.HIGH, trace("H"));
				state(A).when(always()).transition(B, always(), Priority.LOWEST, trace("L"));
			}
		}.build(machine);

		machine.poll();
		assertEquals("HH", trace.toString());
	}

}

File: kotlin/src/test/kotlin/com/laamella/kode_state_machine/PriorityTests.kt
package com.laamella.kode_state_machine

import com.laamella.kode_state_machine.builder.StateMachineBuilder
import com.laamella.kode_state_machine.builder.stateMachine
import com.laamella.kode_state_machine.priority.Priority
import com.laamella.kode_state_machine.priority.Priority.*
import com.laamella.kode_state_machine.util.SimpleState
import com.laamella.kode_state_machine.util.SimpleState.A
import com.laamella.kode_state_machine.util.SimpleState.B
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

internal class PriorityTests {
    private val trace = StringBuffer()

    private var machineBuilder: StateMachineBuilder<SimpleState, Any, Priority> = stateMachine(NORMAL) {
        state(A) { isAStartState() }
        state(B) { isAnEndState() }
    }

    private fun trace(signature: String): TraceAction {
        return TraceAction(trace, signature)
    }

    @Test
    fun highPrioIsTheOnlyOneFiring() {
        val machine = machineBuilder.more {
            state(A) { transitionsTo(B, priority = HIGH, action = trace("H")) }
            state(A) { transitionsTo(B, priority = NORMAL, action = trace("N")) }
            state(A) { transitionsTo(B, priority = LOWEST, action = trace("L")) }
        }.build()

        machine.poll()
        Assertions.assertEquals("H", trace.toString())
    }

    @Test
    fun normalPriosAreTheOnlyOnesFiringBecauseOtherPrioDoesntMeetCondition() {
        val machine = machineBuilder.more {
            state(A) { transitionsTo(B, condition = never(), priority = HIGH, action = trace("H")) }
            state(A) { transitionsTo(B, condition = never(), action = trace("N")) }
            state(A) { transitionsTo(B, action = trace("N")) }
            state(A) { transitionsTo(B, action = trace("N")) }
        }.build()

        machine.poll()
        Assertions.assertEquals("NN", trace.toString())
    }

    @Test
    fun equalPriosFireTogether() {
        val machine = stateMachine<SimpleState, Any, Priority>(NORMAL) {
            state(A) { transitionsTo(B, priority = HIGH, action = trace("H")) }
            state(A) { transitionsTo(B, priority = HIGH, action = trace("H")) }
            state(A) { transitionsTo(B, priority = LOWEST, action = trace("L")) }
        }.build()

        machine.poll()
        Assertions.assertEquals("HH", trace.toString())
    }
}
